src/app.ts

```diff
@@ -1,26 +1,65 @@
-// example interfaces that can be use
-// TIP: the types mentioned in the interfaces must be fulfilled in order to solve the problem.
-interface TemperatureReading {
-  time: Date
-  temperature: number
-  city: string
-}
-interface TemperatureSummary {
-  first: number
-  last: number
-  high: number
-  low: number
-  average: number
+import { TemperatureReading, TemperatureSummary } from './interfaces'
+
+type ProccesedReadings = {
+  [city: string]: {
+    [date: string]: number[]
+  }
 }
+let processedReadings: ProccesedReadings = {}
 
 export function processReadings(readings: TemperatureReading[]): void {
-  // add here your code
+  const processedData = readings.reduce<ProccesedReadings>((data, value) => {
+    const { city, time, temperature } = value
+    const formatedDate = time.toISOString()
+
+    // Validate if city exists or isn't empty
+    if (city.trim() === '' || !city) {
+      throw new Error('Some register has no valid city')
+    }
+
+    // Create city objects if not exists
+    if (!data[city]) {
+      data[city] = {}
+    }
+
+    // Create date objects if not exists
+    if (!data[city][formatedDate]) {
+      data[city][formatedDate] = []
+    }
+
+    // Push temperatures
+    data[city][formatedDate].push(temperature)
+    return data
+  }, {} as ProccesedReadings)
```

Comment:
Since you're filling the generic type on line 11, we can remove the type assertion:

```suggestion
  }, {})
```

When we should use type assertions and when we shouldn't?

Reply:
We should use type assertions when Typescript compiler can't infer the type, we shouldn't use type assertions if we can define explicitly the type or create a generic type.